File: Interviews/LeetCode/Topic-Wise/2-Pointers/Hard/Minimum-Window-Substring/76.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        vector<int> cnt(58, 0);
        for (char c : t) {
            cnt[c - 'A'] += 1;
        }

        int dist = 1e9, start = -1;
        for (int i = 0, j = 0; j < s.size(); j++) {
            cnt[s[j] - 'A'] -= 1;
            if (cnt[s[j] - 'A'] < 0) {
                while (i < s.size() and cnt[s[i] - 'A'] < 0) {
                    cnt[s[i] - 'A'] += 1;
                    i++;
                }
            }

            bool ok = true;
            for (int k = 0; k < 58; k++) {
                ok &= cnt[k] <= 0;
            }

            if (ok and dist > (j - i + 1)) {
                dist = j - i + 1;
                start = i;
            }
        }

        return dist > s.size() ? "" : s.substr(start, dist);
    }
};

File: Interviews/LeetCode/Topic-Wise/2-Pointers/Hard/Minimum-Window-Substring/76.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        cnt = [0 for i in range(58)]
        for c in t: cnt[ord(c) - 65] += 1
        
        dist, start, i = int(1e9), -1, 0
        for j, ch in enumerate(s):
            cnt[ord(ch) - 65] -= 1
            while i < len(s) and cnt[ord(s[i]) - 65] < 0:
                cnt[ord(s[i]) - 65] += 1
                i += 1
            
            ok = True
            for k in cnt: ok &= k <= 0
            
            if ok and dist > (j - i + 1):
                dist, start = j - i + 1, i
        
        return "" if dist > len(s) else s[start : start + dist]